File: abc012/d.py
INF = 1000000000
N, M = map(int, input().split())
mem = [[INF for i in range(N)] for j in range(N)]

for i in range(M):
    a, b, t = map(int, input().split())
    mem[a - 1][b - 1] = mem[b - 1][a - 1] = t

al = INF
an = -1
for s in range(N):
    ml = 0
    done = [False] * N
    l = [INF] * N
    l[s] = 0
    while True:
        d = -1
        for i in range(N):
            if (d == -1 or l[i] < l[d]) and not done[i]:
                d = i
        if d == -1:
            break
        done[d] = True
        for i in range(N):
            l[i] = min(l[i], l[d] + mem[d][i])
        ml = max(ml, l[d])
    if al > ml:
        an = s
        al = ml
print(al)

File: abc012/d.cpp
#include<iostream>
#include<algorithm>
#define INF 1000000000
using namespace std;


int N, M, mem[300][300], l[300], done[300], al = INF, an, mx, d;
int main()
{
  for(int i = 0; i < 300; i++)
    for(int j = 0; j < 300; j++)
      mem[i][j] = INF;
  cin >> N >> M;
  for(int i = 0; i < M; i++)
  {
    int a, b, t;
    cin >> a >> b >> t;
    a--; b--;
    mem[a][b] = mem[b][a] = t;
  }

  for(int s = 0; s < N; s++)
  {
    for(int i = 0; i < 300; i++)
    {
      done[i] = false;
      l[i] = INF;
    }
    l[s] = 0;
    mx = 0;
    while(true)
    {
      d = -1;
      for(int i = 0; i < N; i++)
        if((d == -1 || l[i] < l[d]) && !done[i])
          d = i;
      if(d == -1)
        break;
      done[d] = true;
      for(int i = 0; i < N; i++)
        l[i] = min(l[i], l[d] + mem[d][i]);
      mx = max(mx, l[d]);
    }
    if(mx < al)
    {
      al = mx;
      an = s;
    }
  }
  cout << al << endl;
}
